metabonding/src/claim.rs

```diff
@@ -35,44 +41,72 @@ pub trait ClaimModule:
     /// user_lkmex_staked_amount: BigUint,
     /// signature: 120 bytes
     #[endpoint(claimRewards)]
-    fn claim_rewards(&self, claim_args: MultiValueEncoded<ClaimArgPair<Self::Api>>) {
+    fn claim_rewards(&self, raw_claim_args: MultiValueEncoded<ClaimArgPair<Self::Api>>) {
         require!(self.not_paused(), "May not claim rewards while paused");
         require!(
-            claim_args.raw_len() / CLAIM_NR_ARGS_PER_PAIR <= MAX_CLAIM_ARG_PAIRS,
+            raw_claim_args.raw_len() / CLAIM_NR_ARGS_PER_PAIR <= MAX_CLAIM_ARG_PAIRS,
             "Too many arguments"
         );
 
         let caller = self.blockchain().get_caller();
         let current_week = self.get_current_week();
+        let args = self.validate_and_collect_claim_args(&caller, current_week, raw_claim_args);
+        let rewards = self.claim_all_from_args(current_week, args);
+        if !rewards.is_empty() {
+            self.send().direct_multi(&caller, &rewards);
+        }
+    }
+
+    fn validate_and_collect_claim_args(
+        &self,
+        caller: &ManagedAddress,
+        current_week: Week,
+        raw_args: MultiValueEncoded<ClaimArgPair<Self::Api>>,
+    ) -> ClaimArgArray<Self::Api> {
         let last_checkpoint_week = self.get_last_checkpoint_week();
-        let rewards_nr_first_grace_weeks = self.rewards_nr_first_grace_weeks().get();
+        let nr_first_grace_weeks = self.rewards_nr_first_grace_weeks().get();
 
-        let mut args = ArrayVec::<ClaimArgsWrapper<Self::Api>, MAX_CLAIM_ARG_PAIRS>::new();
-        for arg in claim_args {
+        let mut grace_weeks_progress =
+            self.get_grace_weeks_progress(caller, nr_first_grace_weeks, current_week);
+        let mut shifting_progress = self.get_shifting_progress(caller, current_week);
+        let mut args = ClaimArgArray::new();
+        for raw_arg in raw_args {
             let (week, user_delegation_amount, user_lkmex_staked_amount, signature) =
-                arg.into_tuple();
+                raw_arg.into_tuple();
 
             require!(
-                !self.rewards_claimed(&caller, week).get(),
-                "Already claimed rewards for this week"
-            );
-            require!(week <= last_checkpoint_week, "No checkpoint for week yet");
-            require!(
-                self.is_claim_in_time(week, current_week, rewards_nr_first_grace_weeks),
-                "Claiming too late"
+                week >= FIRST_WEEK && week <= last_checkpoint_week,
+                INVALID_WEEK_NR_ERR_MSG
             );
 
-            let checkpoint: RewardsCheckpoint<Self::Api> = self.rewards_checkpoints().get(week);
+            if grace_weeks_progress.is_week_valid(week) {
+                require!(
+                    grace_weeks_progress.can_claim_for_week(week),
+                    ALREADY_CLAIMED_ERR_MSG
+                );
+
+                grace_weeks_progress.set_claimed_for_week(week);
```

Comment:
Side-effects in a function that does not suggest this would happen (like also mentioned below)